library/ZendService/Twitter/Media.php

```diff
@@ -0,0 +1,251 @@
+        $this->data['media_id']       = 0;
+        $this->data['segment_index']  = 0;
+        $this->data['media_type']     = $media_type;
+
+        if (null !== $imageFilename) {
+            $this->data['image_filename'] = $imageFilename;
+        }
+    }
+
+    /**
+     * Main call into the upload workflow
+     *
+     * @param  Client $httpClient
+     * @return Twitter\Response
+     * @throws Exception If the file can't be opened.
+     */
+    public function upload(Client $httpClient)
+    {
+        if (! $this->validateFile($this->data['image_filename'])) {
+            throw new Exception\InvalidMediaException('Failed to open ' . $this->data['image_filename']);
+        }
+
+        if (empty($this->data['media_type'])) {
+            throw new Exception\InvalidMediaException('No Media Type given.');
+        }
+
+        $params = [];
+        $httpClient->setUri(self::UPLOAD_BASE . self::UPLOAD_PATH);
+
+        $params['total_bytes'] = filesize($this->data['image_filename']);
+        $params['media_type']  = $this->data['media_type'];
+
+        $response = $this->initUpload($httpClient, $params);
+        $initResponse = $response->toValue();
+
+        if (! $response->isSuccess()) {
+            throw new Exception\MediaUploadException('Failed to init the upload with Twitter.');
+        }
+
+        $this->data['media_id'] = $initResponse->media_id;
+
+        $this->appendUpload($httpClient, $params);
+
+        return $this->finalizeUpload($httpClient, $params);
+    }
+
+    /**
+     * Initalize the upload with Twitter
+     *
+     * @param Http\Client $httpClient
+     * @params array $params
+     * @return Twitter\Response
+     */
+    protected function initUpload($httpClient, $params)
+    {
+        $payload = [
+            'command'     => 'INIT',
+            'media_type'  => $params['media_type'],
+            'total_bytes' => $params['total_bytes'],
+        ];
+
+        $httpClient->resetParameters();
+        $httpClient->setHeaders(['Content-type' => 'application/x-www-form-urlencoded']);
+        $httpClient->setMethod('POST');
+        $httpClient->setParameterPost($payload);
+        $response = $httpClient->send();
+        return new Response($response);
+    }
+
+    /**
+     * Send the chunks of the file to Twitter
+     *
+     * @param Http\Client $httpClient
+     * @params array $params
+     * @return boolean
+     */
+    protected function appendUpload($httpClient, $params)
+    {
+        $payload = [
+            'command'  => 'APPEND',
+            'media_id' => $this->data['media_id'],
+        ];
+
+        set_error_handler($this->createErrorHandler(), E_WARNING);
+        $fileHandle = fopen($this->data['image_filename'], 'rb');
+        restore_error_handler();
+
+        if (! $fileHandle) {
+            throw new Exception\MediaUploadException('Failed to open the file in the APPEND method.');
+        }
+
+        while (! feof($fileHandle)) {
```

Comment:
It _is_ buffering, though. We can only send `$chunkSize` bytes at a time to Twitter (limits they set), so we read that many from the file at a time, sending them in an `APPEND` operation to the API, until the entire file has been read. For large files, being able to do so also saves memory for the process performing the upload, as the entire file is never in memory at one time.